pygmt/clib/session.py

```diff
@@ -1528,21 +1533,29 @@ def virtualfile_from_data(
         ...
         <vector memory>: N = 3 <7/9> <4/6> <1/3>
         """
-        kind = data_kind(data, x, y, z, required_z=required_z)
-
-        if check_kind == "raster" and kind not in ("file", "grid"):
-            raise GMTInvalidInput(f"Unrecognized data type for grid: {type(data)}")
-        if check_kind == "vector" and kind not in (
-            "file",
-            "matrix",
-            "vectors",
-            "geojson",
-        ):
-            raise GMTInvalidInput(f"Unrecognized data type for vector: {type(data)}")
+        kind = data_kind(
+            data, x, y, z, required_z=required_z, required_data=required_data
+        )
+
+        if check_kind == "raster":
+            valid_kinds = ("file", "grid")
+        elif check_kind == "vector":
+            valid_kinds = ("file", "matrix", "vectors", "geojson")
```

Comment:
Suggested refactoring. Makes it a little clearer that raster includes "grid", and vector includes "matrix", "vectors", "geojson".

```suggestion
        valid_kinds = ("file", "arg") if required_data is False else ("file",)
        if check_kind == "raster":
            valid_kinds += ("grid",)
        elif check_kind == "vector":
            valid_kinds += ("matrix", "vectors", "geojson")
```